subsys/net/lib/lwm2m/lwm2m_rw_cbor.c

```diff
@@ -0,0 +1,554 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME net_lwm2m_cbor
+#define LOG_LEVEL CONFIG_LWM2M_LOG_LEVEL
+
+#include <logging/log.h>
+#include <sys/util.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+
+#include <net/lwm2m.h>
+#include "lwm2m_object.h"
+#include "lwm2m_rw_cbor.h"
+#include "lwm2m_engine.h"
+#include "lwm2m_util.h"
+
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#define CPKT_CBOR_W_SZ(pos, cpkt) ((size_t)(pos) - (size_t)(cpkt)->data - (size_t)(cpkt)->offset)
+
+#define ICTX_CBOR_R_SZ(pos, ictx) ((size_t)pos - (size_t)(ictx)->in_cpkt->data - (ictx)->offset)
+
+static int put_time(struct lwm2m_output_context *out, struct lwm2m_obj_path *path, int64_t value)
+{
+	/* CBOR time output format is unspecified but SenML CBOR uses string format.
+	 * Let's stick into the same format with plain CBOR
+	 */
+	struct tm dt;
+	char time_str[sizeof("1970-01-01T00:00:00-00:00")] = { 0 };
+	int len;
+	int str_sz;
+	int tag_sz;
+	bool ret;
+
+	if (gmtime_r((time_t *)&value, &dt) != &dt) {
+		LOG_ERR("unable to convert from secs since Epoch to a date/time construct");
+		return -EINVAL;
+	}
+
+	/* Time construct to a string. Time in UTC, offset to local time not known */
+	len = snprintk(time_str, sizeof(time_str),
+			"%04d-%02d-%02dT%02d:%02d:%02d-00:00",
+			dt.tm_year+1900,
+			dt.tm_mon+1,
+			dt.tm_mday,
+			dt.tm_hour,
+			dt.tm_min,
+			dt.tm_sec);
+
+	if (len < 0 || len > sizeof(time_str) - 1) {
+		LOG_ERR("unable to form a date/time string");
+		return -EINVAL;
+	}
+
+	ZCBOR_STATE_E(states, 0, CPKT_BUF_W_PTR(out->out_cpkt), CPKT_BUF_W_SIZE(out->out_cpkt),  1);
+
+	/* TODO: Are tags required? V1.1 leaves this unspecified but some servers require tags */
```

Comment:
Again, it's unclear to me what is the actual leftover work here.

Reply:
Removed TODO, left the rest of the comment.